### ISSUE TITLE
ArrayIndexOutOfBoundsException in FastList

### ISSUE THREAD

**Issue by antonsarov** [brettwooldridge/HikariCP#634]
During normal operations we got these exceptions (using HikariCP 2.4.3 with Hibernate)

```
java.lang.ArrayIndexOutOfBoundsException: -1
	at com.zaxxer.hikari.util.FastList.remove(FastList.java:134)
	at com.zaxxer.hikari.pool.ProxyConnection.untrackStatement(ProxyConnection.java:168)
	at com.zaxxer.hikari.pool.ProxyStatement.close(ProxyStatement.java:73)
	at org.hibernate.jdbc.AbstractBatcher.closePreparedStatement(AbstractBatcher.java:534)
	at org.hibernate.jdbc.AbstractBatcher.closeStatement(AbstractBatcher.java:269)
	at org.hibernate.jdbc.AbstractBatcher.closeQueryStatement(AbstractBatcher.java:285)
	at org.hibernate.jdbc.AbstractBatcher.closeQueryStatement(AbstractBatcher.java:212)
	at org.hibernate.loader.Loader.doQuery(Loader.java:726)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:236)
	at org.hibernate.loader.Loader.doList(Loader.java:2213)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2104)
	at org.hibernate.loader.Loader.list(Loader.java:2099)
	at org.hibernate.loader.hql.QueryLoader.list(QueryLoader.java:378)
	at org.hibernate.hql.ast.QueryTranslatorImpl.list(QueryTranslatorImpl.java:338)
	at org.hibernate.engine.query.HQLQueryPlan.performList(HQLQueryPlan.java:172)
	at org.hibernate.impl.SessionImpl.list(SessionImpl.java:1121)
	at org.hibernate.impl.QueryImpl.list(QueryImpl.java:79)
```
and

```
java.lang.ArrayIndexOutOfBoundsException: -1
	at com.zaxxer.hikari.util.FastList.add(FastList.java:85)
	at com.zaxxer.hikari.pool.ProxyConnection.trackStatement(ProxyConnection.java:191)
	at com.zaxxer.hikari.pool.ProxyConnection.prepareStatement(ProxyConnection.java:310)
	at com.zaxxer.hikari.pool.HikariProxyConnection.prepareStatement(HikariProxyConnection.java)
	at org.hibernate.jdbc.AbstractBatcher.getPreparedStatement(AbstractBatcher.java:505)
	at org.hibernate.jdbc.AbstractBatcher.getPreparedStatement(AbstractBatcher.java:423)
	at org.hibernate.jdbc.AbstractBatcher.prepareQueryStatement(AbstractBatcher.java:139)
	at org.hibernate.loader.Loader.prepareQueryStatement(Loader.java:1547)
	at org.hibernate.loader.Loader.doQuery(Loader.java:673)
	at org.hibernate.loader.Loader.doQueryAndInitializeNonLazyCollections(Loader.java:236)
	at org.hibernate.loader.Loader.doList(Loader.java:2213)
	at org.hibernate.loader.Loader.listIgnoreQueryCache(Loader.java:2104)
	at org.hibernate.loader.Loader.list(Loader.java:2099)
	at org.hibernate.loader.criteria.CriteriaLoader.list(CriteriaLoader.java:94)
	at org.hibernate.impl.SessionImpl.list(SessionImpl.java:1569)
	at org.hibernate.impl.CriteriaImpl.list(CriteriaImpl.java:283)
```
Is there some concurrency happening here which causes the problem?

Looking at the changelog I did not find anything `FastList`-related in the newer versions.

**Comment by nitincchauhan** [comment]
@antonsarov from changelog, Issue [591](https://github.com/brettwooldridge/HikariCP/issues/591) is fixed in 2.4.5. check if this issue is similar.

**Comment by brettwooldridge** [comment]
@antonsarov 

@nitincchauhan is correct, this issue is likely fixed in 2.4.5.
